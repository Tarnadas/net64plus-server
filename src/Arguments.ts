import { ArgumentParser } from 'argparse'

import * as fs from 'fs'
import * as path from 'path'

import { Settings, DEFAULT_SETTINGS } from './models/Settings.model'

export class Arguments {
  public settings: Settings

  private readonly parser = new ArgumentParser({
    addHelp: true,
    description: 'Net64+ server'
  })

  constructor () {
    let settings: Settings | undefined
    const settingsPath = path.join(__dirname, '../settings.json')
    try {
      settings = JSON.parse(fs.readFileSync(settingsPath, {
        encoding: 'utf8'
      }))
    } catch (err) {
      fs.writeFileSync(settingsPath, JSON.stringify(DEFAULT_SETTINGS, undefined, 2))
      console.info('Failed to find or parse settings.json file. Using default settings instead and created a settings.json just for you.')
    }

    this.parser.addArgument(['--port', '-P'], {
      type: (int: string) => parseInt(int)
    })
    this.parser.addArgument(['--gamemode', '-g'], {
      type: (int: string) => parseInt(int)
    })
    this.parser.addArgument(['--disableGamemodeVote', '-G'], {
      action: 'storeConst',
      constant: false,
      dest: 'enableGamemodeVote'
    })
    this.parser.addArgument(['--passwordRequired', '-pr'], {
      action: 'storeConst',
      constant: true
    })
    this.parser.addArgument(['--password', '-p'])
    this.parser.addArgument(['--name', '-n'])
    this.parser.addArgument(['--domain', '-D'])
    this.parser.addArgument(['--description', '-d'])
    this.parser.addArgument(['--enableWebHook', '-w'], {
      action: 'storeConst',
      constant: true
    })
    this.parser.addArgument(['--apiKey', '-k'])
    this.parser.addArgument(['--verbose', '-v'], {
      action: 'storeConst',
      constant: true
    })
    this.parser.addArgument(['--skipPortCheck', '-s'], {
      action: 'storeConst',
      constant: true
    })
    const parsed = this.parser.parseArgs() as Settings

    this.settings = {} as any
    /* eslint-disable @typescript-eslint/ban-ts-ignore */
    Object.entries(DEFAULT_SETTINGS).forEach(([key, defaultValue]: [ string, any ]) => {
      // @ts-ignore
      this.settings[key] = parsed[key]
      // @ts-ignore
      if (this.settings[key] == null) this.settings[key] = (settings ? settings[key] : null)
      // @ts-ignore
      if (this.settings[key] == null) this.settings[key] = defaultValue
    })
  }
}
