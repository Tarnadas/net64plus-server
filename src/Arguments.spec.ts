import * as fs from 'fs'

import { Arguments } from './Arguments'

jest.mock('fs')

describe('Arguments', () => {
  let args: Arguments

  beforeEach(() => {
    process.argv = process.argv.slice(0, 2)
    console.info = jest.fn()
  })

  it('should be created', () => {
    args = new Arguments()

    expect(args).toBeDefined()
  })

  it('should match snapshot', () => {
    args = new Arguments()

    expect(args.settings).toMatchSnapshot()
  })

  describe('with command line options', () => {
    it('should set number', () => {
      process.argv = process.argv.concat('--port', '1337')

      args = new Arguments()

      expect(args.settings.port).toEqual(1337)
    })

    it('should set string', () => {
      process.argv = process.argv.concat('--name', 'server')

      args = new Arguments()

      expect(args.settings.name).toEqual('server')
    })

    it('should set boolean', () => {
      process.argv = process.argv.concat('--disableGamemodeVote')

      args = new Arguments()

      expect(args.settings.enableGamemodeVote).toBe(false)
    })
  })

  describe('with JSON options', () => {
    it('should set number', () => {
      const mockSettings = JSON.stringify({
        port: 1337
      })
      jest.spyOn(fs, 'readFileSync').mockImplementationOnce(() => mockSettings)

      args = new Arguments()

      expect(args.settings.port).toEqual(1337)
    })

    it('should set string', () => {
      const mockSettings = JSON.stringify({
        name: 'server'
      })
      jest.spyOn(fs, 'readFileSync').mockImplementationOnce(() => mockSettings)

      args = new Arguments()

      expect(args.settings.name).toEqual('server')
    })

    it('should set boolean', () => {
      const mockSettings = JSON.stringify({
        enableGamemodeVote: false
      })
      jest.spyOn(fs, 'readFileSync').mockImplementationOnce(() => mockSettings)

      args = new Arguments()

      expect(args.settings.enableGamemodeVote).toBe(false)
    })
  })
})
