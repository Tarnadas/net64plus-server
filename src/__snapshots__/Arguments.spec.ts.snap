// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arguments should match snapshot 1`] = `
Object {
  "apiKey": "",
  "description": "The **best** Net64+ server ever

:unicorn_face:",
  "domain": "",
  "enableGamemodeVote": true,
  "enableWebHook": false,
  "gamemode": 1,
  "name": "A Net64+ Server",
  "password": "",
  "passwordRequired": false,
  "port": 3678,
  "skipPortCheck": false,
  "verbose": false,
}
`;
